### ISSUE TITLE
CSVLoader.loadDataSource() cannot load data without objective variable

### ISSUE THREAD

**Issue by k-tamura** [oracle/tribuo#53]
**problem**

`CSVLoader.loadDataSource()` cannot load  data which does not have objective (response) variable. For example:

```java
// Load Titanic test data downloaded from Kaggle (https://www.kaggle.com/c/titanic)
ListDataSource dataource = csvLoader.loadDataSource(Paths.get("test.csv"),"Survived");
List<Prediction> predicts = model.predict(dataource);
```

this code results:

```
Exception in thread "main" java.lang.IllegalArgumentException: Response Survived not found in file file:/home/tamura/git/tribuo-examples/titanic/test_removed.csv
    at org.tribuo.data.csv.CSVLoader.validateResponseNames(CSVLoader.java:286)
    at org.tribuo.data.csv.CSVLoader.innerLoadFromCSV(CSVLoader.java:244)
    at org.tribuo.data.csv.CSVLoader.loadDataSource(CSVLoader.java:238)
    at org.tribuo.data.csv.CSVLoader.loadDataSource(CSVLoader.java:209)
    at org.tribuo.data.csv.CSVLoader.loadDataSource(CSVLoader.java:184)
    at org.tribuo.data.csv.CSVLoader.loadDataSource(CSVLoader.java:138)
    at TitanicSurvivalClassifier.main(TitanicSurvivalClassifier.java:74)
```

It's a little inconvenient.

**Solution**

Add methods to load data without response variable name.

**Comment by Craigacp** [comment]
The `CSVLoader` is designed for purely numerical data, not the mixture of numerical and text data that comprises the Kaggle Titanic dataset, so this will probably error out later on anyway. For mixed numerical and text data you should use `CSVDataSource` and construct a `RowProcessor` which allows full control of how the features are extracted from the columnar data. `CSVDataSource` has a boolean `outputRequired` flag which if set to false will generate `Example`s which contain the unknown output for that output type (e.g. `Label.UNKNOWN` for classification tasks) if the output doesn't exist in the csv file.

However you're right that there should be a way to load a numerical CSV which doesn't have a response. We'll fix that.

**Comment by k-tamura** [comment]
@Craigacp Thank you for your comment. I wrote the following code first:

```java
com.nhl.dflib.DataFrame df = com.nhl.dflib.csv.Csv.loader().load("titanic/train.csv");
System.out.println(df.head(5));
com.nhl.dflib.DataFrame selectedDataFrame = df.selectColumns("Survived", "Pclass", "SibSp", "Parch", "Fare");
com.nhl.dflib.csv.Csv.save(selectedDataFrame, "titanic/train_removed.csv");

LabelFactory labelFactory = new LabelFactory();
CSVLoader csvLoader = new CSVLoader<>(',',labelFactory);

ListDataSource dataource = csvLoader.loadDataSource(Paths.get("titanic/train_removed.csv"),"Survived");
TrainTestSplitter dataSplitter = new TrainTestSplitter<>(dataource, 0.7, 1L);

MutableDataset trainingDataset = new MutableDataset<>(dataSplitter.getTrain());
MutableDataset testingDataset = new MutableDataset<>(dataSplitter.getTest());

Trainer<Label> trainer = new LogisticRegressionTrainer();
Model<Label> model = trainer.train(trainingDataset);

LabelEvaluator evaluator = new LabelEvaluator();
LabelEvaluation evaluation = evaluator.evaluate(model, testingDataset);
System.out.println(evaluation);
```

Based on your advice, I made the following changes:

```java
Tokenizer tokenizer = new BreakIteratorTokenizer(Locale.US);
LabelFactory labelFactory = new LabelFactory();
ResponseProcessor<Label> responseProcessor = new FieldResponseProcessor<>("Survived","0",labelFactory);
Map<String, FieldProcessor> fieldProcessors = new HashMap<>();
fieldProcessors.put("PassengerId",new IdentityProcessor("PassengerId"));
fieldProcessors.put("Survived",new DoubleFieldProcessor("Survived"));
fieldProcessors.put("Pclass",new DoubleFieldProcessor("Pclass"));
fieldProcessors.put("Name",new TextFieldProcessor("Name",new BasicPipeline(tokenizer,2)));
fieldProcessors.put("Sex",new TextFieldProcessor("Sex",new BasicPipeline(tokenizer,2)));
fieldProcessors.put("Age",new DoubleFieldProcessor("Age"));
fieldProcessors.put("SibSp",new DoubleFieldProcessor("SibSp"));
fieldProcessors.put("Parch",new DoubleFieldProcessor("Parch"));
fieldProcessors.put("Ticket",new TextFieldProcessor("Ticket",new BasicPipeline(tokenizer,2)));
fieldProcessors.put("Fare",new DoubleFieldProcessor("Fare"));
fieldProcessors.put("Cabin",new TextFieldProcessor("Cabin",new BasicPipeline(tokenizer,2)));
fieldProcessors.put("Embarked",new TextFieldProcessor("Embarked",new BasicPipeline(tokenizer,2)));
RowProcessor<Label> rp = new RowProcessor<>(responseProcessor,fieldProcessors);
Path path = Paths.get("titanic/train.csv");
CSVDataSource<Label> source = new CSVDataSource<>(path,rp,true);

TrainTestSplitter dataSplitter = new TrainTestSplitter<>(source, 0.7, 1L);

MutableDataset<Label> trainingDataset = new MutableDataset<>(dataSplitter.getTrain());
MutableDataset<Label> testingDataset = new MutableDataset<>(dataSplitter.getTest());

Trainer<Label> trainer = new LogisticRegressionTrainer();
Model<Label> model = trainer.train(trainingDataset);

LabelEvaluator evaluator = new LabelEvaluator();
LabelEvaluation evaluation = evaluator.evaluate(model, testingDataset);
System.out.println(evaluation);
```
I can see that the text column was automatically One-Hot encoded and get the very high score.

**Comment by k-tamura** [comment]
Thank you for your comment. I wrote this code first:

```java
com.nhl.dflib.DataFrame df = com.nhl.dflib.csv.Csv.loader().load("titanic/train.csv");
System.out.println(df.head(5));
com.nhl.dflib.DataFrame selectedDataFrame = df.selectColumns("Survived", "Pclass", "SibSp", "Parch", "Fare");
com.nhl.dflib.csv.Csv.save(selectedDataFrame, "titanic/train_removed.csv");

LabelFactory labelFactory = new LabelFactory();
CSVLoader csvLoader = new CSVLoader<>(',',labelFactory);

ListDataSource dataource = csvLoader.loadDataSource(Paths.get("titanic/train_removed.csv"),"Survived");
TrainTestSplitter dataSplitter = new TrainTestSplitter<>(dataource, 0.7, 1L);

MutableDataset trainingDataset = new MutableDataset<>(dataSplitter.getTrain());
MutableDataset testingDataset = new MutableDataset<>(dataSplitter.getTest());

Trainer<Label> trainer = new LogisticRegressionTrainer();
Model<Label> model = trainer.train(trainingDataset);

LabelEvaluator evaluator = new LabelEvaluator();
LabelEvaluation evaluation = evaluator.evaluate(model, testingDataset);
System.out.println(evaluation);
```

After getting your comment, I changed as follows and got the same result:

```java
Tokenizer tokenizer = new BreakIteratorTokenizer(Locale.US);
LabelFactory labelFactory = new LabelFactory();
ResponseProcessor<Label> responseProcessor = new FieldResponseProcessor<>("Survived","0",labelFactory);
Map<String, FieldProcessor> fieldProcessors = new HashMap<>();
fieldProcessors.put("PassengerId", new IdentityProcessor("PassengerId"));
fieldProcessors.put("Pclass", new DoubleFieldProcessor("Pclass"));
fieldProcessors.put("SibSp", new DoubleFieldProcessor("SibSp"));
fieldProcessors.put("Parch", new DoubleFieldProcessor("Parch"));
fieldProcessors.put("Fare", new DoubleFieldProcessor("Fare"));
RowProcessor<Label> rp = new RowProcessor<>(responseProcessor,fieldProcessors);
Path path = Paths.get("titanic/train.csv");
CSVDataSource<Label> source = new CSVDataSource<>(path,rp,true);

TrainTestSplitter dataSplitter = new TrainTestSplitter<>(source, 0.7, 1L);

MutableDataset<Label> trainingDataset = new MutableDataset<>(dataSplitter.getTrain());
MutableDataset<Label> testingDataset = new MutableDataset<>(dataSplitter.getTest());

Trainer<Label> trainer = new LogisticRegressionTrainer();
Model<Label> model = trainer.train(trainingDataset);

LabelEvaluator evaluator = new LabelEvaluator();
LabelEvaluation evaluation = evaluator.evaluate(model, testingDataset);
System.out.println(evaluation);
```

BTW, I added the following code but it does not work as I intended:

```java
// Text column
fieldProcessors.put("Name", new TextFieldProcessor("Name", new BasicPipeline(tokenizer,2)));
// Including null column
fieldProcessors.put("Age", new DoubleFieldProcessor("Age"));
```

I'm looking for how to change text to double value, fill null value ​​and run one-hot-encoding but I did not find the methods. Should I create custom classes?

**Comment by Craigacp** [comment]
I'm not clear exactly what you mean by convert text to 0/1. Do you want a single categorical feature for the full text field (i.e. "Braund, Mr. Owen Harris"=1) or features of the form "Name@2-N=Leslie"=1 for each word or bigram in the text field. The basic pipeline will do the latter. In general as Tribuo operates on sparse features it tends to blow out categoricals into a one hot encoding unless there is special processing.

I made a couple of modifications to the processing pipeline you suggested, reducing the number of tokens in the "Sex" column as it's only one word (though you could equally use an `IdentityProcessor` here instead of the `BasicPipeline`), and changing the "PassengerId" field so it's stored as example metadata rather than a feature. Using it as a feature was causing the logistic regression to overfit to it. The null age value is implicitly represented as a zero, because Tribuo treats everything as sparse data and an empty value in `DoubleFieldProcessor` is treated as a non-present feature. To insert a specific value for an empty field you'd have to write a custom implementation of `FieldProcessor` or modify the current `DoubleFieldProcessor`. We're happy to accept contributions to build out the columnar processing infrastructure or to add new features to classes like `DoubleFieldProcessor`.

```java
ResponseProcessor<Label> responseProcessor = new FieldResponseProcessor<>("Survived","0",labelFactory);
Map<String, FieldProcessor> fieldProcessors = new HashMap<>();
fieldProcessors.put("Pclass", new DoubleFieldProcessor("Pclass"));
fieldProcessors.put("SibSp", new DoubleFieldProcessor("SibSp"));
fieldProcessors.put("Parch", new DoubleFieldProcessor("Parch"));
fieldProcessors.put("Fare", new DoubleFieldProcessor("Fare"));
fieldProcessors.put("Sex",new TextFieldProcessor("Sex",new BasicPipeline(tokenizer,1)));
fieldProcessors.put("Age", new DoubleFieldProcessor("Age"));
fieldProcessors.put("Name", new TextFieldProcessor("Name", new BasicPipeline(tokenizer,2)));
IntExtractor passengerID = new IntExtractor("PassengerId");
RowProcessor<Label> rp = new RowProcessor<>(Collections.singletonList(passengerID),null,responseProcessor,fieldProcessors,Collections.emptySet());
```

You should be able to supply that same `RowProcessor` to a `CSVDataSource` and set the output required flag to false to load the test data in.

You can inspect the feature domain Tribuo extracted by looking at the `FeatureMap` returned by `trainingDataset.getFeatureMap()` or `model.getFeatureIDMap()`.

**Comment by k-tamura** [comment]
Thank you for for your kindness.

> I'm not clear exactly what you mean by convert text to 0/1.

Sorry I didn't explain it clearly enough. Firstly, I tried to replace:

 - Sex: `male` -> `0`, `female`: `1`
 - Age: null -> Median value
 - Embarked: `C`, `S`, `K` -> One-hot-encoding

but I could not find good way. Shouldn't we use Tribuo to do data preprocessing?

**Comment by Craigacp** [comment]
Tribuo's data preprocessing set is currently a bit limited when compared to something like pandas, but you can do most of what you suggest already. 
- For Sex the `IdentityProcessor` will emit a one hot encoding of the sex, so you'll get features with name "Sex#male" and value 1.0, and "Sex#female" with value 1.0 depending on the field value. Tribuo's examples are implicitly sparse, so you don't need to add the other features which are zero, that's assumed. The `RegexFieldProcessor` could be used to emit a "Sex#MATCHES_ALL" feature with value 1.0, for either male or female depending on the regex used, which would then give you a 1.0 for one value and a 0.0 otherwise (but it's a little confusing to look at). Again due to the implicit sparse-ness you won't actually see the features with a 0.0 value, but it's computed when necessary by all the training algorithms.
- For "Age", currently there is no way to add the median value to replace a null. You could put in a sentinel value and then use a feature transformer to convert the sentinel into the median. That would require writing a slightly modified `DoubleFieldProcessor` that could emit the sentinel value, and then writing a median transformer that could replace the sentinel value with the median. The transform package is [here](https://github.com/oracle/tribuo/tree/main/Core/src/main/java/org/tribuo/transform/transformations), but it doesn't have a feature replacer (it's not tricky to add, we just haven't had need for it yet). The reason we can't do this in one pass is that the `RowProcessor` is stateless and only operates on a single row at a time. Only the feature transformers can see the whole dataset.
- For "Embarked" Tribuo's `IdentityProcessor` will emit features with name = "<column-name>#<column-value>" and value = 1.0, so it will do exactly what you want, a one hot encoding as the values which aren't present are implicitly zero.

**Comment by k-tamura** [comment]
```java
fieldProcessors.put("Embarked",new IdentityProcessor("Embarked"));
```
This line generates feature map:

- `Embarked@`
- `Embarked@S`
- `Embarked@C`
- `Embarked@Q`

Can Tribuo replace null with 's' to avoid generating `Embarked@`?

**Comment by Craigacp** [comment]
Not directly without modifying the `IdentityProcessor`. You could do that, or you could implement a very short [FeatureProcessor](https://github.com/oracle/tribuo/blob/main/Data/src/main/java/org/tribuo/data/columnar/FeatureProcessor.java) which replaces one feature with another. Feature processors are run on the feature list produced from a given row allowing the user to modify the features before they get put into the `Example`. The latter is probably something we should add to Tribuo anyway, the `FeatureProcessor` mechanism is fairly powerful but we only have a single implementation in the Github repo.

**Comment by k-tamura** [comment]
Thank you! I will firstly implement a `FeatureProcessor`.
